library/core/src/array/mod.rs

```diff
@@ -426,6 +427,18 @@ impl<T: Clone, const N: usize> Clone for [T; N] {
     }
 }
 
+#[unstable(feature = "random", issue = "130703")]
+impl<T: Random, const N: usize> Random for [T; N] {
+    fn random(source: &mut (impl RandomSource + ?Sized)) -> Self {
+        let mut buf = [const { MaybeUninit::uninit() }; N];
+        for elem in &mut buf {
+            elem.write(T::random(source));
```

Comment:
If `T::random` panics the elements already present in `buf` won't be freed.

We generally try to avoid leaking memory when possible, we generally do that with a guard type that frees on `Drop`, and is `mem::forgot` when everything is fine.

We already have such a [`Guard` type](https://github.com/rust-lang/rust/blob/d2f335d58e6c346f94910d0f49baf185028b44be/library/core/src/array/mod.rs#L870-L886) for uninit-arrays, see [here](https://github.com/rust-lang/rust/blob/d2f335d58e6c346f94910d0f49baf185028b44be/library/core/src/array/mod.rs#L857-L866) for how to use it.